src/main/scala/vectorpipe/VectorPipe.scala

```diff
@@ -0,0 +1,104 @@
+import vectorpipe.vectortile.export._
+
+import geotrellis.proj4.{CRS, LatLng, WebMercator}
+import geotrellis.spark.SpatialKey
+import geotrellis.spark.tiling.{ZoomedLayoutScheme, LayoutLevel}
+import geotrellis.vector.Geometry
+import geotrellis.vectortile._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+import org.locationtech.jts.{geom => jts}
+
+object VectorPipe {
+
+  case class Options(
+    maxZoom: Int,                    // Largest (most resolute) zoom level to generate
+    minZoom: Option[Int],            // Smallest (least resolute) zoom level to generate
+    srcCRS: CRS,                     // projection of the original geometry
+    destCRS: Option[CRS],            // projection for generated vectortiles
+    tileResolution: Int = 4096       // "pixel" resolution of output vectortiles
+  )
+  object Options {
+    def forZoom(zoom: Int) = Options(zoom, None, LatLng, None)
+    def forAllZoomsFrom(zoom: Int) = Options(zoom, Some(0), LatLng, None)
+    def forAllZoomsWithSrcProjection(zoom: Int, crs: CRS) = Options(zoom, Some(0), crs, None)
+  }
+
+  def apply(input: DataFrame, pipeline: vectortile.Pipeline, layerName: String, options: Options): Unit = {
+    val geomColumn = pipeline.geometryColumn
+    assert(input.columns.contains(geomColumn), s"Input DataFrame must contain a column `${geomColumn}` of JTS Geometry")
+    // TODO check the type of the geometry column
+
+    val srcCRS = options.srcCRS
+    val destCRS = options.destCRS.getOrElse(WebMercator)
+    val maxZoom = options.maxZoom
+    val minZoom = math.min(math.max(0, options.minZoom.getOrElse(options.maxZoom)), options.maxZoom)
+    val zls = ZoomedLayoutScheme(destCRS, options.tileResolution)
+
+    // Reproject geometries if needed
+    val reprojected = input.withColumn(geomColumn,
+                                       st_reprojectGeom(col(geomColumn),
+                                                        lit(srcCRS.toProj4String),
+                                                        lit(destCRS.toProj4String)))
+
+    // Prefilter data for first iteration and key geometries to initial layout
+    val keyColumn = {
+      var prepend = "_"
+      while (input.columns.contains(prepend ++ "keys")) { prepend = "_" ++ prepend }
+      prepend ++ "keys"
+    }
+
+    def reduceKeys = udf { arr: Array[SpatialKey] => arr.map{ case SpatialKey(x, y) => SpatialKey(x / 2, y / 2) }.toSet.toArray }
+
+    def generateVectorTiles[G <: Geometry](df: DataFrame, level: LayoutLevel): RDD[(SpatialKey, VectorTile)] = {
+      val zoom = level.zoom
+      val clip = udf { (g: jts.Geometry, k: SpatialKey) => pipeline.clip(g, k, level) }
+
+      pipeline
+        .select(df, zoom)
+        .withColumn(keyColumn, explode(col(keyColumn)))
+        .repartition(col(keyColumn)) // spread copies of possibly ill-tempered geometries around cluster
+        .withColumn(geomColumn, clip(col(geomColumn)))
```

Comment:
When I attempted to use this branch to run a pipeline, I got a `cannot cast to type scala.function1` error. 

I fixed with this diff:
```diff
diff --git a/src/main/scala/vectorpipe/VectorPipe.scala b/src/main/scala/vectorpipe/VectorPipe.scala
index bab14f2..0d0c467 100644
--- a/src/main/scala/vectorpipe/VectorPipe.scala
+++ b/src/main/scala/vectorpipe/VectorPipe.scala
@@ -62,7 +62,7 @@ object VectorPipe {
         .select(df, zoom)
         .withColumn(keyColumn, explode(col(keyColumn)))
         .repartition(col(keyColumn)) // spread copies of possibly ill-tempered geometries around cluster
-        .withColumn(geomColumn, clip(col(geomColumn)))
+        .withColumn(geomColumn, clip(col(geomColumn), col(keyColumn)))
         .rdd
         .map { r => (r.getAs[SpatialKey](keyColumn), pipeline.pack(r, zoom)) }
         .groupByKey
```

Reply:
The PR was in an un-debugged state at the point you tried it.  I've had a test actually run with the changes below.